V2: training with cross validation

hwpang opened the pull request:
## Description
This PR intends to resolve some parts of https://github.com/chemprop/chemprop/issues/524 and https://github.com/chemprop/chemprop/issues/581.

- [x] Fix the issue where the same list of indices repeat num_folds times
- [x] The `data_split` function is not doing cross-validation correctly. See https://github.com/chemprop/chemprop/issues/609. Needs it to be fixed to complete this PR.
- [x] Add support to use cross validation in the training script
- [x] Add support to use ensembling in the training script
- [x] Add training with cross validation in cli test

Below are some relevant discussions:
- Output directory structure for cross-validation and ensembling: https://github.com/chemprop/chemprop/issues/581#issuecomment-1896322966 (may need to consult https://github.com/chemprop/chemprop/issues/591, and https://github.com/chemprop/chemprop/pull/599 for the output structure and location)

davidegraff commented:
can we break this up into two PRs: (i) ensembling and then (ii) cross-validation? The former will fundamentally change the object model of the train and predict CLIs while the latter will _sometimes_ change it? These are both decently large features and should be reviewed independently.

hwpang commented:
@davidegraff Thanks for the comment! Yeah that makes sense. I will split up this PR.

hwpang commented:
@davidegraff I have removed the changes relevant to ensembling from this PR, and this PR should now only contain cross-validation related efforts. I moved the ensembling related efforts to https://github.com/chemprop/chemprop/pull/631. The ensembling efforts depend on this PR, so I think it's easier to review and merge this one in, and then move onto https://github.com/chemprop/chemprop/pull/631.

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -542,170 +531,220 @@ def main(args):
         raise ArgumentError(
-            argument=None, message="'val_path' must be specified if 'test_path' is provided!"
+            argument=None, message="'test_path' must be specified if 'val_path' is provided!"
         )  # TODO: In v1 this wasn't the case?
```
This TODO might need to be broken out into a separate issue?

hwpang replied:
Thanks! I made this issue to keep track of this TODO: https://github.com/chemprop/chemprop/issues/639.

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -542,170 +531,220 @@ def main(args):
+                logger.info(f"Test results: {results}")
+
+            p_model = args.output_dir / "model.pt"
+            input_scalers = []  # TODO: we should add descriptor scalers here
```
Again, consider breaking this TODO into an issue

JacksonBurns replied:
(and removing the inline comment)

hwpang replied:
Added this issue, will remove the inline comment: https://github.com/chemprop/chemprop/issues/640

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -542,170 +531,220 @@ def main(args):
+                    )
+                    for i in range(n_components)
+                ]
+                if args.mpn_shared:
+                    mp_block = MulticomponentMessagePassing(
+                        mp_blocks[0], n_components, args.mpn_shared
+                    )
+                else:
+                    mp_block = MulticomponentMessagePassing(
+                        mp_blocks, n_components, args.mpn_shared
+                    )
```
```
mp_block = MulticomponentMessagePassing(
    mp_blocks[0] if args.mpn_shared else mp_blocks, n_components, args.mpn_shared
)
```

davidegraff replied:
Is this `if/else` necessary? It seems like it's taken care of inside [`MulticomponentMessagePassing.__init__`](https://github.com/chemprop/chemprop/blob/327c31e932db8cea92e9d9a221c288eaeee911cb/chemprop/nn/message_passing/multi.py#L50):
```python
self.blocks = nn.ModuleList([blocks[0]] * self.n_components if shared else blocks)
```

hwpang replied:
@davidegraff Good call! You're right. I will remove this `if/else`.

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -542,170 +531,220 @@ def main(args):
+                if len(test_data[0]) > 0:
+                    test_dsets = [make_dataset(data, args.rxn_mode) for data in test_data[fold_idx]]
+                    test_dset = data.MulticomponentDataset(test_dsets)
+                    test_loader = data.MolGraphDataLoader(
+                        test_dset, args.batch_size, args.num_workers, shuffle=False
+                    )
+                else:
+                    test_loader = None
```
This can be moved outside of the `if multicomponent:` block right? (and removed from both places in here).

hwpang replied:
Thanks for the suggestion! I move this outside of the `if multicomponent:` block.

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -542,170 +531,220 @@ def main(args):
+
+            tb_logger = TensorBoardLogger(output_dir / "tb_logs")
+            checkpointing = ModelCheckpoint(
+                output_dir / "chkpts",
```
Super nitpick, but I don't see a reason to abbreviate this (even though this is close to the file extension) when `checkpoints` would be more clear.

hwpang replied:
I think this is outside the scope of this PR as I only move this part of code into the for loop and did not change the directory name. We can open a different issue to discuss whether to abbreviate output directories.

KnathanM replied:
I agreed with Jackson and included this in my PR #599

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -542,170 +531,220 @@ def main(args):
+            tb_logger = TensorBoardLogger(output_dir / "tb_logs")
+            checkpointing = ModelCheckpoint(
+                output_dir / "chkpts",
+                "{epoch}-{val_loss:.2f}",
```
Is it possible for this to overwrite previous checkpoints?

davidegraff replied:
yes, provided they have the `output_dir` and same validation loss (to 2 decimal points) on the same epoch

hwpang replied:
With changes implemented in https://github.com/chemprop/chemprop/pull/599, I think `output_dir` would be `$PWD/chemprop_training/<input_stem>/$NOW` so this will not be overwritten.

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -542,170 +531,220 @@ def main(args):
+                save_last=True,
+            )
+
+            early_stopping = EarlyStopping("val_loss", patience=5, mode=monitor_mode)
```
RE: our discussion during the hackathon last week. v1 would never actually _stop_ training, it would just load the best model from all epochs, acting as if it had stopped there. This will be different, in that v2 will actually _stop training_. Should we make this _very_ clear somewhere? Is this intended?

davidegraff replied:
We should make this configurable. I think I just added this in when I originally scaffolded out the training CLI from other code I've written. Long term, it would be nice if we could have a separate config file for callbacks, but for now, we can just add a `--patience` arg and set the default to `None` to indicate training for the full number of epochs:
```python
patience = args.patience if args.patience is not None else args.epochs
```

hwpang replied:
Thanks for the suggestions! I have added a `--patience` arg and set the default to `None` as suggested.

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -542,170 +531,220 @@ def main(args):
+                logger=tb_logger,
+                enable_progress_bar=True,
+                accelerator="auto",
+                devices=args.n_gpu if torch.cuda.is_available() else 1,
```
There's a bit of difficulty in this - lightning has a lot of ways to use `devices`, see: https://lightning.ai/docs/pytorch/stable/accelerators/gpu_basic.html#:~:text=The%20table%20below%20lists%20examples%20of%20possible%20input%20formats%20and%20how%20they%20are%20interpreted%20by%20Lightning.

We should follow up about this offline. The way we pass this arg should probably change if we want to expose all the functionality of `devices`.

davidegraff replied:
Agreed. IMO we should have an `int` arg `--devices` that's just the _number_ of devices to train on. If a user wants to limit the GPU set, they can just run it like so:
```
$ CUDA_VISIBLE_DEVICES=0,2 chemprop train --devices -1
```
or we could also use the `find_usable_cuda_devices` method so that a user doesn't even need to bother with limiting GPU visibility

JacksonBurns reviewed (commented):
Some minor details, but some larger questions for discussion. Overall, this looks quite good.

davidegraff commented on chemprop/data/splitting.py:
```diff
@@ -87,7 +87,7 @@ def split_data(
+            if not (2 <= num_folds <= len(datapoints)):
                 raise ValueError(
                     f"Number of folds for cross-validation must be between 3 and {max_folds} (length of data) inclusive for `--split cv` and 2 and {max_folds} (length of data) inclusive for `--split cv_no_val` (got {num_folds})."
                 )
```
```suggestion
                min_folds = 2 if SplitType.get(split) == SplitType.CV_NO_VAL else 3
                raise ValueError(
                    f"invalid number of folds requested! got: {num_folds}, but expected between {min_folds} and {len(datapoints)} (i.e., number of datapoints), inclusive, for split type: {repr(split)}"
                )
```

davidegraff commented on chemprop/cli/train.py:
```diff
@@ -458,6 +458,12 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         default=0,
         help="Seed for PyTorch randomness (e.g., random initial weights).",
     )
+    parser.add_argument(
+        "--patience",
+        type=int,
+        default=None,
+        help="Number of epochs to wait for improvement before early stopping.",
+    )
```
Note: this can stay because we've discussed this already, but stuff like this should be a separate PR in the future (there really is no minimum size!)

hwpang replied:
Thanks for the suggestion! Will keep this change here and open up new PRs in the future.

davidegraff commented:
> Just so I'm clear on the logical structure here, you've restructured the train CLI to effectively treat _all_ training as having `k` separate splits and training one model for each of those splits. It's at the end of model training that you then return and check whether this model was trained as part of CV or not, and if so, structure the output directory as `OUTPUT/FOLD_K/...`. Otherwise, you structure it normally? I just want to make sure I understand this before I approve. Thanks!

@hwpang can you check if this is right?

hwpang commented:
@davidegraff Yes that is correct!

davidegraff commented on chemprop/cli/train.py:
```diff
@@ -691,63 +701,83 @@ def main(args):
+        test_datas = [test_data]
     else:
-        test_loader = None
+        train_datas = train_data
+        val_datas = val_data
+        test_datas = test_data
```
```suggestion
        splits = (train_data, val_data, test_data)
```

davidegraff commented on chemprop/cli/train.py:
```diff
@@ -691,63 +701,83 @@ def main(args):
-        )
+    no_cv = args.num_folds == 1
+    if no_cv:
+        train_datas = [train_data]
+        val_datas = [val_data]
+        test_datas = [test_data]
```
```suggestion
        splits = ([train_data], [val_data], [test_data])
```

davidegraff commented on chemprop/cli/train.py:
```diff
@@ -691,63 +701,83 @@ def main(args):
 
-    model = build_model(args, train_dset)
-    logger.info(model)
+    for fold_idx, (train_data, val_data, test_data) in enumerate(zip(train_datas, val_datas, test_datas)):
```
```suggestion
    for fold_idx, (train_data, val_data, test_data) in enumerate(splits):
```

davidegraff reviewed (approved):
minor style suggestions, otherwise lgtm

hwpang commented:
All tests pass! I believe this is now ready to merge.
